pubsub/gcloud/aio/pubsub/subscriber_client.py

```diff
@@ -132,60 +166,75 @@ def create_subscription(self,
-                      ) -> StreamingPullFuture:
+        async def pull(self, subscription: str, max_messages: int,
+                       *, session: Optional[Session] = None) -> List[SubscriberMessage]:
             """
-            Create subscription through pubsub client, scheduling callbacks
-            on the event loop.
+            Pull messages from subscription
             """
-            sub_keepalive = self._subscriber.subscribe(
-                subscription,
-                self._wrap_callback(callback),
-                flow_control=flow_control)
-
-            self.loop.add_signal_handler(signal.SIGTERM, sub_keepalive.cancel)
-
-            return sub_keepalive
-
-        def run_forever(self, sub_keepalive: StreamingPullFuture) -> None:
+            url = f'{API_ROOT}/v1/{subscription}:pull'
+            headers = await self._headers()
+            body = {
+                'maxMessages': max_messages,
+            }
+            s = session or self.session
+            resp = await s.post(url, data=json.dumps(body).encode(), headers=headers)
+            resp = await resp.json()
+            return [SubscriberMessage.from_api_dict(m) for m in resp['receivedMessages']]
+
+        async def acknowledge(self, subscription: str, ack_ids: List[str],
+                              *, session: Optional[Session] = None) -> None:
+            """
+            Acknowledge messages by ackIds
+            """
+            url = f'{API_ROOT}/v1/{subscription}:acknowledge'
+            headers = await self._headers()
+            body = {
+                'ackIds': ack_ids,
+            }
+            s = session or self.session
+            resp = await s.post(url, data=json.dumps(body).encode(), headers=headers)
+            return
+
+        async def modify_ack_deadline(self, subscription: str, ack_ids: List[str],
+                                      ack_deadline_seconds: int,
+                                      *, session: Optional[Session] = None) -> None:
             """
-            Start the asyncio loop, running until it is either SIGTERM-ed or
-            killed by keyboard interrupt. The StreamingPullFuture parameter is
-            used to cancel subscription in the case that an unexpected
-            exception is thrown. You can also directly pass the `.subscribe()`
-            method call instead like so:
-                sub.run_forever(sub.subscribe(callback))
+            Modify messages' ack deadline. Set ack deadline to 0 to nack messages
             """
+            url = f'{API_ROOT}/v1/{subscription}:modifyAckDeadline'
+            headers = await self._headers()
+            body = {
+                'ackIds': ack_ids,
+                'ackDeadlineSeconds': ack_deadline_seconds,
+            }
+            s = session or self.session
+            resp = await s.post(url, data=json.dumps(body).encode(), headers=headers)
+            return
+
+
+        async def subscribe(self,
```

Comment:
I'm not sure we need to worry about creating this method -- I think the goal here should be mostly to implement _just the API interface_ and leave the actual scheduling outside of the domain of this library. I think our main goal should be avoiding re-inventing the wheel here -- the exact threading models used in clients of this library are not something we can predict and each app will probably have different requirements; rather let them implement those cases and/or offload this to scheduling libraries.

We simply need to provide the primitives, not the higher-level operations.

If we do want to provide this sort of higher-level interface _as well_, it would probably be best to separate it out entirely from the API code -- perhaps a `SubscriptionManager` class or such. And perhaps only available in the `-aio-` lib, since it makes no sense in the case of a single thread? That said, I think we could just as easily toss this sort of thing in the docs for now and worry about creating the full interface only once we've explored other options and seen what it is we actually end up needing downstream and what we can offload to a more Broadway-like lib.

Reply:
Yeah, that makes sense. My reasoning is that we'll have to have this interface distributed to all our services, so I need to put it _somewhere_. For the purpose of starting a discussion, that seemed like an ok place :)

Reply:
Makes sense! I'm curious how many of our apps will need the same code here and what we might need to customize -- for example, being able to customize the size of the `asyncio.Queue` is probably very important to clients (as an aside, check out https://nullprogram.com/blog/2020/05/24/ for an awesome read on some gotchas here), though maybe that's just a matter of passing in `max_messages` to the constructor?

Reply:
wow that's a great article. I now feel a little better for using queues at the first place. I'll be playing with `max_message` and what not as first load tests reveals extreme inefficiency of this solution. Feel free to toss any ideas of realization, I'll test them out.